lib/email/sender.rb

```diff
@@ -514,5 +474,144 @@ def self.bounceable_reply_address?
     def self.bounce_address(bounce_key)
       SiteSetting.reply_by_email_address.sub("%{reply_key}", "verp-#{bounce_key}")
     end
+
+    def add_identification_field_headers(topic, post)
+      # If the topic was created from an incoming email, then the Message-ID from
+      # that email will be the canonical reference, otherwise the canonical reference
+      # will be <topic/TOPIC_ID@host>. The canonical reference is used in the
+      # References header.
+      #
+      # This is so the sender of the original email still gets their nice threading
+      # maintained (because their mail client will initiate threading based on
+      # the Message-ID it generated) in the case where there is an incoming email.
+      #
+      # In the latter case, everyone will start their thread with the canonical reference,
+      # because we send it in the References header for all emails.
+      topic_canonical_reference_id = Email::MessageIdService.generate_for_topic(
+        topic, canonical: true, use_incoming_email_if_present: true
+      )
+
+      # Whenever we reply to a post directly _or_ quote a post, a PostReply
+      # record is made, with the reply_post_id referencing the newly created
+      # post, and the post_id referencing the post that was quoted or replied to.
+      referenced_posts = Post.includes(:incoming_email)
+        .joins("INNER JOIN post_replies ON post_replies.post_id = posts.id ")
+        .where("post_replies.reply_post_id = ?", post.id)
+        .order(id: :desc)
+
+      referenced_post_message_ids = referenced_posts.map do |referenced_post|
+        if referenced_post.incoming_email&.message_id.present?
+          "<#{referenced_post.incoming_email.message_id}>"
+        else
+          if referenced_post.post_number == 1
+            topic_canonical_reference_id
+          else
+            Email::MessageIdService.generate_for_post(referenced_post)
+          end
+        end
+      end
+
+      # See https://www.ietf.org/rfc/rfc2822.txt for the message format
+      # specification, more useful information can be found in Email::MessageIdService
+      #
+      # The References header is how mail clients handle threading. The Message-ID
+      # must always be unique.
+      if post.post_number == 1
+        @message.header['Message-ID']  = Email::MessageIdService.generate_for_topic(topic)
+        @message.header['References']  = [topic_canonical_reference_id]
```

Comment:
I'm still unconvinced here. Is the topic somehow distinct from the first post? in that is has a meaningful message body of its own? I'm still of the opinion that there should be no `References` header at all for post 1, just as there's no `In-Reply-To` either.